flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/RuntimeMonitorHandler.java

```diff
@@ -44,27 +47,40 @@
 public class RuntimeMonitorHandler extends SimpleChannelInboundHandler<Routed> {
 	
 	private static final Charset ENCODING = Charset.forName("UTF-8");
+
+	private final JobManagerArchiveRetriever retriever;
 	
 	private final RequestHandler handler;
 	
 	private final String contentType;
 	
-	public RuntimeMonitorHandler(RequestHandler handler) {
-		if (handler == null) {
-			throw new NullPointerException();
-		}
-		this.handler = handler;
+	public RuntimeMonitorHandler(RequestHandler handler, JobManagerArchiveRetriever retriever) {
+		this.retriever = checkNotNull(retriever);
+		this.handler = checkNotNull(handler);
 		this.contentType = (handler instanceof RequestHandler.JsonResponse) ? "application/json" : "text/plain";
 	}
 	
 	@Override
 	protected void channelRead0(ChannelHandlerContext ctx, Routed routed) throws Exception {
+		// Redirect to leader if necessary
+		String redirectAddress = retriever.getRedirectAddress();
```

Comment:
Get rid of this `getRedirectAddress` method here and simply ask for the current leader. Then compare the leader address with the local job manager address to decide whether you have to redirect or not.